### DISCUSSION TITLE
How to add an icon for certain extensions?

### DISCUSSION THREAD

**Discussion by Fenriuz** [kyazdani42/nvim-web-devicons#541] [Q&A]
How to add an icon for certain extensions?

I don't really know how to formulate my question, but with devicons I could do this `let g: WebDevIconsUnicodeDecorateFileNodesPatternSymbols ['. * Test. * \. Js $'] = 'ﭧ'`
And the result was the following:
![image](https://user-images.githubusercontent.com/63080658/104346293-dfccdb80-54c4-11eb-8f81-787d6d16eee0.png)

How can I achieve the same result with this plugin?

**Comment by kyazdani42** [comment]
Hi, regexes are not supported at the moment, i think this could be added as a feature if it doesn't impact performance.

**Comment by Fenriuz** [comment]
I get it, it would be great if that feature was added.

**Comment by kyazdani42** [comment]
I agree, ill see when I have some time to implement this

**Comment by poetaman** [comment]
@kyazdani42 Is it possible to add a icon for a extension like ``.vimrc`` or ``.zshrc``? Currently, they do not get applied icon/color of ``vim`` or ``zsh`` for some reason. Note: this is simpler than regex, it is hardcoded extension name.

**Comment by kyazdani42** [comment]
i'm not sure what's your question here @reportaman ? could you be a little bit more specific please ?

**Comment by luissimas** [comment]
Hi, I'm not sure if this is exactly what @Fenriuz means, but I was looking for a similar behaviour while using [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua), and in the process of trying to implement it I may have found something that works.

Before:
![image](https://user-images.githubusercontent.com/41878330/120079169-1462bd80-c089-11eb-92d7-8cde7c713f90.png)

After:
![image](https://user-images.githubusercontent.com/41878330/120079181-23497000-c089-11eb-9566-2cb5721b7aa3.png)

I couldn't find a good way of implementing it in nvim-web-devicons itself, the way that the `get_icon` function works seems fine, so I actually changed something in how [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua) handles the file extensions.

@kyazdani42, I see that [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua) uses the `vim.fn.fnamemodify` function with the `":e"` modifier to get the file extension and then pass it to nvim-web-devicons, I think that this way of getting the file extension doesn't work very well when we have file extensions with multiple dots (like `foo.test.js`).

What I did was change the `vim.fn.fnamemodify` on [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua) with a lua pattern match that returns the full string after the first `.` occurence: `string.match(name, "%.(.*)")`. Now I can define the icons like this in nvim-web-devicons's setup function:
```lua
require("nvim-web-devicons").setup {
  override = {
    ["test.js"] = {
      icon = "ﭧ",
      color = "#cbcb41",
      name = "JavascriptTest"
    }
  },
  default = true
}
```

The problem is that I don't know if this will break anything in [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua), so I would like to ask @kyazdani42's opinion on this.

This is far from being the regex-like feature, but I think it covers some use cases that are very common.

If it's ok I can submit a PR to [nvim-tree](https://github.com/kyazdani42/nvim-tree.lua) and then add some icons and doc regarding this behaviour to this repo later.

**Comment by kyazdani42** [comment]
hi @luissimas, send a PR and i'll do a set of personnal test to ensure it's working as expected. It might break icon display for certain filenames but i'm not sure as i haven't seen your implementation yet :)

**Comment by Looooopy** [comment]
regexes would be a great feature. 👍 

Like it is now we have to 

- Rename a couple of files to match this naming convention to get nice icons 

- Stick with default icons and keep our own naming convention. 

I suppose I will stick with our naming for now until we get regexes in this library.

Great pull-request from @luissimas 🥇  that matches all dots in the end of the string.

Keep up the good work @kyazdani42 💯 and thanks for bringing us this library.

**Comment by qRoC** [comment]
For test files you can do:

```lua
local devicons = require "nvim-web-devicons"

local function get_ext(name)
  if name:find "^.+%.test%..+$" or name:find "^.+Test%..+$" or name:find "^.+_test%..+$" then
    return "test"
  end

  if name:find "^.+%.spec%..+$" or name:find "^.+Spec%..+$" or name:find "^.+_spec%..+$" then
    return "spec"
  end

  return name:match "^.*%.(.*)$" or ""
end

local get_icon = devicons.get_icon
devicons.get_icon = function(name, ext, opts)
  return get_icon(name, ext or get_ext(name), opts)
end

local get_icon_colors = devicons.get_icon_colors
devicons.get_icon_colors = function(name, ext, opts)
  return get_icon_colors(name, ext or get_ext(name), opts)
end

devicons.setup {
  override = {
    ["test"] = { icon = "ﭧ", color = "#cbcb41", name = "TestFile" },
    ["spec"] = { icon = "ﭧ", color = "#cbcb41", name = "SpecFile" },
  },
}
```

Example:
```txt
HelloTest.php
hello.test.js
hello_test.go
```

**Comment by y0nei** [comment]
Can this also be done for files that contain "env" in their name? Or generally user defined regex filters?
Eg.
```
.env-sample
.env_example
.env.production
```

**Comment by alex-courtis** [comment]
Unfortunately we don't have regex support.

You can add your own specific icons via API: https://github.com/nvim-tree/nvim-web-devicons#set-an-icon

**Comment by davmacario** [comment]
I had a similar issue but with files of the type `.gitlab-ci.rules.yml`, and ended up using a solution based on what @qRoC suggested:

```lua
local function get_special_ext(name)
    if name:find(".*%.gitlab%-ci.*%.yml") then -- Match <>.gitlab-ci<>.yml
         return "gitlab-ci.yml" -- Return `gitlab-ci.yml` as the extension
    end
    return nil
end

local devicons = require("nvim-web-devicons")
local get_icon = devicons.get_icon
devicons.get_icon = function(name, ext, opts)
      return get_icon(name, get_special_ext(name) or ext, opts)
end
local get_icon_colors = devicons.get_icon_colors
devicons.get_icon_colors = function(name, ext, opts)
      return get_icon_colors(name, get_special_ext(name) or ext, opts)
end
```

The thing is, files like `.gitlab-ci.<something>.yml` are actually assigned a non-default icon (they are correctly recognized as YAML files), so I had to swap the order of `get_special_ext(name)` and `ext` when calling `get_icon` and `get_icon_colors`.

This also requires that, by default, `get_special_ext` returns `nil` unless one of the rules I define works, so I can use the function to define only the special matching rules.

**Comment by davmacario** [comment]
Forgot to add - I also defined a "custom extension" (`gitlab-ci.yaml`) when calling `require("nvim-web-devicons").setup()` so that I can also catch files with the name `<something>.gitlab-ci.yml`. Then, I can just return `gitlab-ci.yml` as the extension to get the correct icon.

```lua
devicons.setup({
    -- [...]
    override_by_extension = {
	["gitlab-ci.yml"] = <my custom icon>
    },
})

```

**Comment by Yash-xoxo** [comment]
### Understanding the Icon System

The codebase appears to use a mapping system where file extensions are associated with specific icons. Based on the code snippets shown, you'll need to:

### Implementation Steps

**1. Add your icon to the icon set:**
```javascript
// In the icons definition file
const icons = {
  // ... existing icons
  'your-extension': {
    icon: 'icon-name',
    color: '#hexcolor'
  }
}
```

**2. Map your file extension(s):**
```javascript
// In the extension mapping
const fileExtensions = {
  'rs': 'rust',
  'toml': 'config',
  'your-ext': 'your-extension',  // Add your mapping here
  // ... other mappings
}
```

**3. Consider multiple extensions:**
If your language/tool uses multiple file extensions:
```javascript
const fileExtensions = {
  'jsx': 'react',
  'tsx': 'react',  // Multiple extensions → same icon
  'ext1': 'your-extension',
  'ext2': 'your-extension',
}
```

### Best Practices

**Icon Selection:**
- Choose an icon that's visually distinct from existing ones
- Ensure color contrast works in both light/dark themes
- Consider using the official logo/brand colors if applicable

**Testing:**
Test your icon with various scenarios:
```
- Single file: `file.ext`
- Dotfiles: `.extensionrc`
- Multiple dots: `file.config.ext`
- Case sensitivity: `file.EXT` vs `file.ext`
```

**Documentation:**
Update the icon reference documentation to include:
- Which extensions map to this icon
- The icon's purpose/language
- Any special cases (e.g., config files vs source files)

### Example PR Structure

1. **Add icon definition** (icons.ts or similar)
2. **Map file extensions** (fileExtensions.ts)
3. **Add tests** if the project has them
4. **Update docs** (README or icon reference)
5. **Include screenshot** showing the icon in use

### Common Gotchas

⚠️ **Extension conflicts:** Check if your extension is already mapped
⚠️ **Case sensitivity:** Some filesystems are case-insensitive
⚠️ **Priority:** If multiple rules could match, which takes precedence?
⚠️ **Regex patterns:** Some icons may use patterns, not just extensions

### Finding the Right Files

Look for these files in the repository:
```
src/icons/
src/extensions/
src/mappings/
src/fileIcons.ts
src/iconDefinitions.ts
```
